src/internal/pfsdb/commits.go

```diff
@@ -0,0 +1,956 @@
+		return nil, nil, errors.Wrap(err, "getting children commits")
+		// if children is missing, assume commit is HEAD of some branch.
+	}
+	return commitParentRows, commitChildrenRows, nil
+}
+
+func getCommitRowByCommitKey(ctx context.Context, tx *pachsql.Tx, commit *pfs.Commit) (*commitRow, error) {
+	row := &commitRow{}
+	if commit == nil {
+		return nil, ErrCommitMissingInfo{Field: "Commit"}
+	}
+	id := CommitKey(commit)
+	err := tx.QueryRowxContext(ctx, fmt.Sprintf("%s WHERE commit_id=$1", getCommit), id).StructScan(row)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrCommitNotFound{CommitID: id}
+		}
+		return nil, errors.Wrap(err, "scanning commit row")
+	}
+	return row, nil
+}
+
+func parseCommitInfoFromRow(row *commitRow) (*pfs.CommitInfo, error) {
+	commit, err := parseCommitFromRow(row)
+	if err != nil {
+		return nil, errors.Wrap(err, "getting commit from row")
+	}
+	commitInfo := &pfs.CommitInfo{
+		Commit:      commit,
+		Origin:      &pfs.CommitOrigin{Kind: pfs.OriginKind(pfs.OriginKind_value[strings.ToUpper(row.Origin)])},
+		Started:     timeToTimestamppb(row.StartTime),
+		Finishing:   timeToTimestamppb(row.FinishingTime),
+		Finished:    timeToTimestamppb(row.FinishedTime),
+		Description: row.Description,
+		Details: &pfs.CommitInfo_Details{
+			CompactingTime: bigIntToDurationpb(row.CompactingTime),
+			ValidatingTime: bigIntToDurationpb(row.ValidatingTime),
+			SizeBytes:      row.Size,
+		},
+	}
+	return commitInfo, nil
+}
+
+func parseCommitFromRow(row *commitRow) (*pfs.Commit, error) {
+	repo := &pfs.Repo{
+		Name: row.RepoName,
+		Type: row.RepoType,
+		Project: &pfs.Project{
+			Name: row.ProjectName,
+		},
+	}
+	parsedId := strings.Split(row.CommitID, "@")
+	if len(parsedId) != 2 {
+		return nil, errors.New(fmt.Sprintf("got invalid commit id from postgres: %s", row.CommitID))
+	}
+	commit := &pfs.Commit{
+		Repo: repo,
+		Id:   parsedId[1],
+		Branch: &pfs.Branch{
+			Repo: repo,
+			Name: row.BranchName.String,
+		},
+	}
+	return commit, nil
+}
+
+// UpsertCommit will attempt to insert a commit. If the commit already exists, it will update its description.
+func UpsertCommit(ctx context.Context, tx *pachsql.Tx, commitInfo *pfs.CommitInfo, opts ...AncestryOpt) error {
+	existingCommit, err := getCommitRowByCommitKey(ctx, tx, commitInfo.Commit)
+	if err != nil {
+		if errors.Is(ErrCommitNotFound{CommitID: CommitKey(commitInfo.Commit)}, errors.Cause(err)) {
+			return CreateCommit(ctx, tx, commitInfo, opts...)
+		}
+		return errors.Wrap(err, "upserting commit")
+	}
+	return UpdateCommit(ctx, tx, existingCommit.RowID, commitInfo, opts...)
+}
+
+// UpdateCommit overwrites an existing commit entry by CommitID.
+func UpdateCommit(ctx context.Context, tx *pachsql.Tx, id CommitID, commitInfo *pfs.CommitInfo, opts ...AncestryOpt) error {
+	if err := validateCommitInfo(commitInfo); err != nil {
+		return err
+	}
+	opt := AncestryOpt{}
+	if len(opts) > 0 {
+		opt = opts[0]
+	}
+	update := commitRow{
```

Comment:
I think we should remove these extra fields that don't go anywhere from the commitRow type.

Reply:
so, the extra field are generally used to derive the `id` of resources that commitInfo depends on like `repo_id` for instance.